src/libsystemd-network/sd-dhcp-lease.c

```diff
@@ -505,6 +505,57 @@ static int lease_parse_domain(const uint8_t *option, size_t len, char **ret) {
         return 0;
 }
 
+static int lease_parse_fqdn(const uint8_t *option, size_t len, char **ret) {
+        _cleanup_free_ char *name = NULL, *normalized = NULL;
+        size_t data_len = len - 3;
+        int r;
+        const uint8_t *data;
+
+        assert(option);
+        assert(ret);
+
+        /* RFC 4702 Section 2
+         *
+         * Byte 0: Flags (S, O, E, N)
+         * Byte 1: RCODE1 (ignored on receipt)
+         * Byte 2: RCODE2 (ignored on receipt)
+         * Bytes 3+: Domain Name */
+
+        if (len <= 3)
+                return 0;
+
+        data = option + 3;
+
+        /* In practice, many servers send DNS wire format regardless of the E flag, so ignore and try wire
+         * format first, then fall back to ASCII if that fails. */
+        r = dns_name_from_wire_format(&data, &data_len, &name);
+        if (r < 0) {
+                /* Wire format failed, try ASCII format */
+                r = dhcp_option_parse_string(option + 3, len - 3, &name);
+                if (r < 0)
+                        return r;
+        }
+
+        if (!name) {
+                *ret = mfree(*ret);
```

Comment:
Please do not do like that. Our coding style does not assume `ret` is initialized.
Why not simply return `-EBADMSG` or so??